packages/gatsby/src/utils/page-data.ts

```diff
@@ -0,0 +1,69 @@
+import fs from "fs-extra"
+import path from "path"
+import { IGatsbyPage } from "../redux/types"
+import { store } from "../redux"
+
+interface IPageDataContext {
+  publicDir: string
+}
+
+type IPagePath = IGatsbyPage["path"]
+
+export const fixedPagePath = (pagePath: IPagePath): IPagePath =>
+  pagePath === `/` ? `index` : pagePath
+
+const getFilePath = (
+  { publicDir }: IPageDataContext,
+  pagePath: IPagePath
+): IPagePath =>
+  path.join(publicDir, `page-data`, fixedPagePath(pagePath), `page-data.json`)
+
+export const read = async (
+  { publicDir }: IPageDataContext,
+  pagePath: IPagePath
+): Promise<IGatsbyPage> => {
+  const filePath = getFilePath({ publicDir }, pagePath)
+  const rawPageData = await fs.readFile(filePath, `utf-8`)
+
+  return JSON.parse(rawPageData)
+}
+
+export const remove = async (
+  { publicDir }: IPageDataContext,
+  pagePath: IPagePath
+): Promise<void> => {
+  const filePath = getFilePath({ publicDir }, pagePath)
+
+  if (fs.existsSync(filePath)) {
+    return await fs.remove(filePath)
+  }
+
+  return Promise.resolve()
+}
+
+export const write = async (
+  { publicDir }: IPageDataContext,
+  { componentChunkName, matchPath, path }: IGatsbyPage,
+  result: any
```

Comment:
Here, I was not really sure what is the shape of the `result` argument so I judge `any` being safer. But feel free to indicate a more precise typing for this line, if existing.